File: algorithms/1038.binary-search-tree-to-greater-sum-tree/solution.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func bstToGst(root *TreeNode) *TreeNode {
	sum := 0
	helper(root, &sum)

	return root
}

func helper(node *TreeNode, sum *int) {
	if node == nil {
		return
	}

	helper(node.Right, sum)

	node.Val += *sum
	*sum = node.Val

	helper(node.Left, sum)
}

File: algorithms/1038.binary-search-tree-to-greater-sum-tree/solution.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::cell::RefCell;
use std::rc::Rc;
impl Solution {
    pub fn bst_to_gst(root: Option<Rc<RefCell<TreeNode>>>) -> Option<Rc<RefCell<TreeNode>>> {
        let mut sum = 0i32;
        Solution::helper(root.as_ref(), &mut sum);

        root
    }

    fn helper(root: Option<&Rc<RefCell<TreeNode>>>, sum: &mut i32) {
        if let Some(node) = root {
            let mut node = node.borrow_mut();

            Solution::helper(node.right.as_ref(), sum);

            node.val += *sum;
            *sum = node.val;

            Solution::helper(node.left.as_ref(), sum);
        }
    }
}
